File: OneWeekend/Python/sence.py
import numpy as np

from typedef import Point3, Color, Hittable, HitRecord, Material
from material import Lambertian, Metal, Dielectric
from utils import random_float, random_vector

class HittableList(Hittable):
    def __init__(self) -> None:
        super().__init__()
        self.list = []

    def clear(self):
        self.list.clear()

    def add(self, object: Hittable):
        self.list.append(object)

    def hit(self, ray, t_min, t_max) -> HitRecord:
        time = t_max
        record = None
        for object in self.list:
            rec = object.hit(ray, t_min, t_max)
            if rec is not None and rec.time < time:
                record = rec
                time = rec.time
        return record

class Sphere(Hittable):
    def __init__(self, center, radius, material):
        super().__init__()
        self.center = center
        self.radius = radius
        self.material = material

    def hit(self, ray, t_min, t_max) -> HitRecord:
        oc = ray.origin - self.center
        dir = ray.direction
        a = dir.length_squared()
        b = 2 * dir.dot(oc)
        c = oc.length_squared() - self.radius ** 2
        dis = b ** 2 - 4 * a * c
        if dis < 0:
            return None
        sqrtd = np.sqrt(dis)
        time = (-b - sqrtd) / (2 * a)
        if time < t_min or time > t_max:
            time = (-b + sqrtd) / (2 * a)
            if time < t_min or time > t_max:
                return None
        point = ray.at(time)
        front_face = True
        normal = (point - self.center) / self.radius
        if ray.direction.dot(normal) > 0:
            normal = -normal
            front_face = False
        return HitRecord(point, normal, time, front_face, self.material)

def random_sense() -> HittableList:
    world  = HittableList()
    ground_material = Lambertian(Color([0.5, 0.5, 0.5]))
    world.add(Sphere(Point3([0,-1000,0]), 1000, ground_material))

    for a in range(-11, 11):
        for b in range(-11, 11):
            choose_mat = random_float(0, 1)
            center = Point3([a + 0.9 * random_float(0, 1), 0.2, b + 0.9 * random_float(0, 1)])
            if ((center - Point3([4, 0.2, 0])).length() > 0.9):
                if (choose_mat < 0.8):
                    albedo = random_vector(0, 1) * random_vector(0, 1)
                    sphere_material = Lambertian(albedo)
                    world.add(Sphere(center, 0.2, sphere_material))
                elif choose_mat < 0.95:
                    albedo = random_vector(0.5, 1)
                    fuzz = random_float(0, 0.5)
                    sphere_material = Metal(albedo, fuzz)
                    world.add(Sphere(center, 0.2, sphere_material))
                else:
                    sphere_material = Dielectric(1.5)
                    world.add(Sphere(center, 0.2, sphere_material))

    material1 = Dielectric(1.5)
    world.add(Sphere(Point3([0, 1, 0]), 1.0, material1))

    material2 = Lambertian(Color([0.4, 0.2, 0.1]))
    world.add(Sphere(Point3([-4, 1, 0]), 1.0, material2))

    material3 = Metal(Color([0.7, 0.6, 0.5]), 0.0)
    world.add(Sphere(Point3([4, 1, 0]), 1.0, material3))

    return world

File: OneWeekend/Rust/src/sence.rs
use std::rc::Rc;

use super::vector::{Vec3, Point};
use super::camera::Ray;
use super::material::Material;
use super::utils::Interval;

pub trait Hittable {
    fn hit(&self, ray: &Ray, interval: Interval) -> Option<HitRecord>;
}

pub struct HitRecord {
    point: Point,
    normal: Vec3,
    front: bool,
    material: Rc<dyn Material>,
    time: f64
}

impl HitRecord {
    fn new(
        point: Point, normal: Vec3, front: bool, material: Rc<dyn Material>, time: f64
    ) -> Self {
        Self { point, normal, front, material, time }
    }

    pub fn normal(&self) -> Vec3 {
        self.normal
    }

    pub fn point(&self) -> Point {
        self.point
    }

    pub fn front(&self) -> bool {
        self.front
    }

    pub fn material(&self) -> Rc<dyn Material> {
        self.material.clone()
    }
}

pub struct Sence {
    objects: Vec<Box<dyn Hittable>>
}

impl Sence {
    pub fn new() -> Self {
        Self { objects: Vec::new() }
    }

    pub fn push(&mut self, hittable: impl Hittable + 'static) {
        self.objects.push(Box::new(hittable));
    }

    pub fn clear(&mut self) {
        self.objects.clear();
    }
}

impl Hittable for Sence {
    fn hit(&self, ray: &Ray, interval: Interval) -> Option<HitRecord> {
        let mut record = None;
        let mut max_time = interval.max();
        for object in &self.objects {
            if let Some(rec) = object.hit(
                ray, Interval::new(interval.min(), max_time)
            ) {
                if rec.time < max_time {
                    max_time = rec.time;
                    record = Some(rec);
                }
            }
        }
        record
    }
}

pub struct Sphere {
    center: Point,
    radius: f64,
    material: Rc<dyn Material>
}

impl Sphere {
    pub fn new(center: Point, radius: f64, material: impl Material + 'static) -> Self {
        Self { center, radius, material: Rc::new(material) }
    }
}

impl Hittable for Sphere {
    fn hit(&self, ray: &Ray, interval: Interval) -> Option<HitRecord> {
        let oc = self.center - ray.origin();
        let a = ray.direction().length_squared();
        let h = ray.direction().dot(&oc);
        let c = oc.length_squared() - self.radius * self.radius;
        let dis = h * h - a * c;
        if dis < 0.0 {
            return None;
        }

        let mut root = (h - libm::sqrt(dis)) / a;
        if !interval.surrounds(root) {
            root = (h + libm::sqrt(dis)) / a;
            if !interval.surrounds(root) {
                return None;
            }
        }
        let time = root;
        let point = ray.at(time);
        let (front, normal) = {
            let n = (point - self.center) / self.radius;
            if n.dot(&ray.direction()) < 0.0 {
                (true, n)
            } else {
                (false, -n)
            }
        };
        Some(HitRecord::new(point, normal, front, self.material.clone(), time))
    }
}
